### PR TITLE
DM-18341: alternate cache discovery

### PR DESCRIPTION
**Pull request by loitly**
https://jira.lsstcorp.org/browse/DM-18341

additional changes:  https://github.com/lsst/suit/pull/14  and https://github.com/IPAC-SW/irsa-ife/pull/62

**Objective:**
The objective is to implement an alternative to multicast cache peer discovery mechanism. Multicast is not supported in the cloud, making Firefly unable to scale horizontally.

The implemented alternative is to use a publish-subscribe messaging system for peer discovery.  The flow is similar to multicast in which it relies on a periodic heartbeat message from each active peer to determine ‘liveliness’.  Subscribing to the same channel determine the membership of the group.

**Implementation Details:**
A custom CachePeerProviderFactory is created to determine which discovery mechanism to use based on configuration.
Messaging is implemented using Redis server and Jedis client.  The preferred deployment of Redis is via Docker.  Jedis dependencies were added to Firefly.  For more details of its usage, look at the attached diagram as well as MessengerTest.

**Tests:**
using PubSub.. to test that it picks up new peer and replicate
https://irsawebdev9.ipac.caltech.edu/dm-18341/suit/admin/status
https://irsawebdev9.ipac.caltech.edu/dm-18341-2/suit/admin/status

using MultiCast... to test existing config still works
https://irsawebdev9.ipac.caltech.edu/dm-18341/irsaviewer/
https://irsawebdev9.ipac.caltech.edu/dm-18341/firefly/

**Unit tests:**
_HINTS:_  to see test’s debug messages, edit ./firefly/config/test/log4j-test.properties, replace log4j.logger.test=WARN with DEBUG

3 tests were created.  
JsonHelperTest is the easiest to test.  Simply run: 
`gradle :firefly:test -x jsTest --tests *JsonHelperTest`

The next 2 tests require a running Redis.  To start one locally..
`docker run --name test-redis -p 6379:6379 -d redis`

To test them:
`gradle :firefly:test -x jsTest --tests *MessengerTest`
`gradle :firefly:test -x jsTest --tests *CachePeerProviderFactoryTest`

To perf/stress test the messaging system..
`gradle -Pcategory=edu.caltech.ipac.TestCategory\$Perf :firefly:test -x jsTest --tests *MessengerTest`

Also: cleanup prop files
    - move mail.prop into common.prop
    - share common.prop with other projects
    - move differences in other projects’ common.prop into <project>.prop

Messenger

### CODE REVIEW

File: src/firefly/java/edu/caltech/ipac/firefly/messaging/Message.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        String subject = helper.getValue("", "header", "subject");
+        Header header = new Header(scope, to, subject);
+        header.setFrom(from);
+        return header;
+    }
+
+    public void setHeader(Scope scope, String to, String subject, String from) {
+        helper.setValue(scope.name(), "header", "scope");
+        if (to != null) helper.setValue(to, "header", "to");
+        if (subject != null) helper.setValue(from, "header", "from");
```
[the remark this answers is not in the document]
Comment (loitly) [reply src/firefly/java/edu/caltech/ipac/firefly/messaging/Message.java]: good catch.  thanks.

File: src/firefly/java/edu/caltech/ipac/firefly/messaging/Messenger.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    /**
+     * Internal handler class used to manage the one to many relationship of Messenger's subscriber and
+     * Jedis's subscriber
+     */
+    static class SubscriberHandle {
+
+        private CopyOnWriteArrayList<Subscriber> subscribers = new CopyOnWriteArrayList<>();
+        ExecutorService executor;
+        private JedisPubSub jPubSub = new JedisPubSub() {
+            public void onMessage(String channel, String message) {
```
[the remark this answers is not in the document]
Comment (loitly) [reply src/firefly/java/edu/caltech/ipac/firefly/messaging/Messenger.java]: Actually, Jedis allows you to subscribe to multiple channels.  That's why channel is sent back onMessage.  In my implementation, you can only subscribe to one 'topic'(channel).  Internally, we combine all common subscribers into one Jedis.subscribe() thread.  This way, we'll only use one thread per defined topic.  In the future, if this become a problem, we can internally combine topics into one Jedis.subscribe() without affecting publicly expose interfaces.  I hope you understand what I mean.

File: src/firefly/java/edu/caltech/ipac/firefly/messaging/Messenger.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+        void addSubscriber(Subscriber sub) {
+            subscribers.add(sub);
+        }
+
+        void removeSubscriber(Subscriber sub) {
+            subscribers.remove(sub);
+            if (subscribers.size() == 0) {
+                executor.shutdown();
+                unsubscribe();
```
[the remark this answers is not in the document]
Comment (loitly) [reply src/firefly/java/edu/caltech/ipac/firefly/messaging/Messenger.java]: You are probably right.  This is not used at the moment, therefore I may missed it.  I'll fix it and add it into tests.

File: src/firefly/java/edu/caltech/ipac/firefly/messaging/Message.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * Date: 2019-03-15
+ *
+ * @author loi
+ * @version $Id: $
+ */
+public class Message {
+    private JsonHelper helper = new JsonHelper();
+
+    public Header getHeader() {
+        Scope scope = Scope.valueOf(helper.getValue(Scope.SELF.name(), "header", "scope"));
```
[the remark this answers is not in the document]
Comment (loitly) [reply src/firefly/java/edu/caltech/ipac/firefly/messaging/Message.java]: This is in anticipation for when we move our messaging over this Messenger.
When using Scope, `to` is the value of Scope.  i.e.  Scope.CHANNEL, `to` is the channel name.

`subject` can be use as a generic way to filter out the messages you want to act on.

File: src/firefly/test/edu/caltech/ipac/util/cache/CachePeerProviderFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    }
+
+//====================================================================
+//  tests for MultiCasts
+//====================================================================
+
+
+    @Test
+    public void multicast_InitialSetup() throws InterruptedException, RemoteException {
+        if (Messenger.isOffline()) return;
```
[the remark this answers is not in the document]
Comment (loitly) [reply src/firefly/test/edu/caltech/ipac/util/cache/CachePeerProviderFactoryTest.java]: No, not at the moment.  However, they can co-exists.

File: src/firefly/java/edu/caltech/ipac/firefly/messaging/Messenger.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (loitly) [reply src/firefly/java/edu/caltech/ipac/firefly/messaging/Messenger.java]: Thanks for the pointer.  It turns out, I was not cleaning up the connections on unsubscribe.  It's fixed now.  I've also added more tests to MessengerTest.

File: src/firefly/java/edu/caltech/ipac/util/cache/CachePeerProviderFactory.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+ * Currently, peerDiscovery may be "PubSub" or "MultiCast".  When not given, "MultiCast" is used.
+ * Similar to MultiCast, multicastGroupPort is used in PubSub to group related caches for replication.
+ *
+ * Date: 2019-04-08
+ *
+ * @author loi
+ * @version $Id: $
+ */
+public class CachePeerProviderFactory extends CacheManagerPeerProviderFactory {
+    static final Logger.LoggerImpl LOG = Logger.getLogger();
```
Comment (robyww) [inline src/firefly/java/edu/caltech/ipac/util/cache/CachePeerProviderFactory.java]: I am impressed. This took some study to implement. I don't know how many people has done it outside of the ehcache people.

File: src/firefly/java/edu/caltech/ipac/firefly/messaging/Message.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ * uses JSON when serialize/deserialize is needed.  However, this is an internal
+ * implementation and it can be changed without breaking Messenger/Message contract.
+ *
+ * Date: 2019-03-15
+ *
+ * @author loi
+ * @version $Id: $
+ */
+public class Message {
+    private JsonHelper helper = new JsonHelper();
```
[the remark this answers is not in the document]
Comment (loitly) [reply src/firefly/java/edu/caltech/ipac/firefly/messaging/Message.java]: I think you misunderstood the intend of Message and Message.Header.  It's designed to be value object and that you would use them to send and receive messages.  The JSON format is an internal implementation.  The parse and toJson is meant to be package protected, but needed to be exposed due to tests and statistics usage.

**Comment by loitly** [comment]
The discovery listener does not recover when `redis` is bounced.  I will fix it and push a new commit.

**Pull request merged**